akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala

```diff
@@ -36,240 +35,135 @@ private object PoolSlot {
 
-    /**
-     * How PoolProcessor changes its `receive`:
-     * waitingExposedPublisher -> waitingForSubscribePending -> unconnected ->
-     * waitingForDemandFromConnection OR waitingEagerlyConnected -> running
-     * Given slot can become get to 'running' state via 'waitingForDemandFromConnection' or 'waitingEagerlyConnected'.
-     * The difference between those two paths is that the first one is lazy - reacts to DispatchCommand and then uses
-     * inport and outport actors to obtain more items.
-     * Where the second one is eager - reacts to SlotShouldConnectCommand from PoolConductor, sends SlotEvent.ConnectedEagerly
-     * back to conductor and then waits for the first DispatchCommand
-     */
-    override def receive = waitingExposedPublisher
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler {
+      private var firstRequest: RequestContext = _
+      private val inflightRequests = new java.util.ArrayDeque[RequestContext]()
 
-    def waitingExposedPublisher: Receive = {
-      case ExposedPublisher(publisher) ⇒
-        exposedPublisher = publisher
-        context.become(waitingForSubscribePending)
-      case other ⇒ throw new IllegalStateException(s"The first message must be `ExposedPublisher` but was [$other]")
-    }
+      private var connectionFlowSource: SubSourceOutlet[HttpRequest] = _
+      private var connectionFlowSink: SubSinkInlet[HttpResponse] = _
 
-    def waitingForSubscribePending: Receive = {
-      case SubscribePending ⇒
-        exposedPublisher.takePendingSubscribers() foreach (s ⇒ self ! ActorPublisher.Internal.Subscribe(s))
-        log.debug("become unconnected, from subscriber pending")
-        context.become(unconnected)
-    }
+      private var isConnected = false
 
-    val unconnected: Receive = {
-      case OnNext(DispatchCommand(rc: RequestContext)) ⇒
-        val (connInport, connOutport) = runnableGraph.run()
-        connOutport ! Request(totalDemand)
-        context.become(waitingForDemandFromConnection(connInport = connInport, connOutport = connOutport, rc))
+      def disconnect(ex: Option[Throwable] = None) = {
+        connectionFlowSource.complete()
+        if (isConnected) {
+          isConnected = false
 
-      case OnNext(ConnectEagerlyCommand) ⇒
-        val (in, out) = runnableGraph.run()
-        onNext(SlotEvent.ConnectedEagerly(slotIx) :: Nil)
-        out ! Request(totalDemand)
-        context.become(waitingEagerlyConnected(connInport = in, connOutport = out))
+          val (inflightRetry, inflightFail) = if (ex.isEmpty) (inflightRequests.asScala, Nil)
+          else inflightRequests.iterator().asScala.partition(_.retriesLeft > 0)
 
-      case Request(_) ⇒ if (remainingRequested == 0) request(1) // ask for first request if necessary
+          val retries = inflightRetry.map(rc ⇒ SlotEvent.RetryRequest(if (rc.retriesLeft > 0) rc.copy(retriesLeft = rc.retriesLeft - 1) else rc)).toList
```

Comment:
> The substream will request one, then when that is sent and a response is read it will request a second.

Yes, that's the flow with persistent HTTP connections. However, if pipelining would work, requests would be dispatched without waiting for the response of a previous request. So, connection closing could also run concurrently with ongoing requests.